lib/builddata.sh

```diff
@@ -1,29 +1,182 @@
+		# Values passed using `--arg` are treated as strings, and so have double quotes added and any JSON
+		# special characters (such as newlines, carriage returns, double quotes, backslashes) are escaped.
+		local jq_args=(--arg value "${value}")
+	else
+		# Values passed using `--argjson` are treated as raw JSON values, and so aren't escaped or quoted.
+		local jq_args=(--argjson value "${value}")
+	fi
+
+	local new_data_file_contents
+	new_data_file_contents="$(jq --arg key "${key}" "${jq_args[@]}" '. + { ($key): ($value) }' "${BUILD_DATA_FILE}")"
+	echo "${new_data_file_contents}" >"${BUILD_DATA_FILE}"
+}
+
+# Retrieve the value of an entry in the build data store from the previous successful build.
+# Returns the empty string if the key wasn't found in the store.
+#
+# Usage:
+# ```
+# build_data::get_previous "python_version"
+# ```
+function build_data::get_previous() {
+	local key="${1}"
+
+	# Older versions of this buildpack used a `key=value` format file instead of JSON,
+	# so we need to support this file format/location too, so older caches can be read.
+	# We check for this file first, so that we correctly handle the case where an app
+	# downgraded and then re-upgraded buildpack version, so has both files in the cache.
+	if [[ -f "${LEGACY_BUILD_DATA_FILE}" ]]; then
+		# The legacy file contains one entry per line, of form `key=value`. Entries were written in an
+		# append-only manner so there could be duplicate entries for each key, so we return only the
+		# last matching entry in the file. The empty string is returned if the key wasn't found.
+		tac "${LEGACY_BUILD_DATA_FILE}" | { grep --perl-regexp --only-matching --max-count=1 "^${key}=\K.*$" || true; }
+	elif [[ -f "${PREVIOUS_BUILD_DATA_FILE}" ]]; then
+		# The `// empty` ensures we return the empty string rather than `null` if the key doesn't exist.
+		jq --raw-output ".${key} // empty" "${PREVIOUS_BUILD_DATA_FILE}"
+	fi
+}
+
+# Retrieve the value of an entry in the build data store from the current build.
+# Returns the empty string if the key wasn't found in the store.
+#
+# Usage:
+# ```
+# build_data::get_current "python_version"
+# ```
+function build_data::get_current() {
+	local key="${1}"
+
+	# Older versions of this buildpack used a `key=value` format file instead of JSON,
+	# so we need to support this file format/location too, so older caches can be read.
+	# We check for this file first, so that we correctly handle the case where an app
+	# downgraded and then re-upgraded buildpack version, so has both files in the cache.
+	if [[ -f "${LEGACY_BUILD_DATA_FILE}" ]]; then
+		# The legacy file contains one entry per line, of form `key=value`. Entries were written in an
+		# append-only manner so there could be duplicate entries for each key, so we return only the
+		# last matching entry in the file. The empty string is returned if the key wasn't found.
+		tac "${LEGACY_BUILD_DATA_FILE}" | { grep --perl-regexp --only-matching --max-count=1 "^${key}=\K.*$" || true; }
```

Comment:
I think this should be removed, since the "current" value can never come from the legacy file, since the currently running build will be writing to the new style file instead?